pkg/reconciler/cache/publishedresources/publishedresources_reconcile_replication.go

```diff
@@ -0,0 +1,146 @@
+	logger.Info("reconciling published resource", "publishedResource", publishedResource.Name)
+
+	gvr := schema.GroupVersionResource{
+		Group:    publishedResource.Spec.Group,
+		Version:  publishedResource.Spec.Version,
+		Resource: publishedResource.Spec.Resource,
+	}
+	cluster := logicalcluster.From(publishedResource)
+
+	clusterName := logicalcluster.From(publishedResource)
+	controllerName := fmt.Sprintf("%s.%s.%s.%s.%s", clusterName, gvr.Version, gvr.Resource, gvr.Group, publishedResource.Name)
+	// TODO: Add locking here when multiple workers are supported.
+	controller := r.controllerRegistry.get(controllerName)
+	// We setup controller even if we are deleting. This is to ensure that we can purge the cache.
+	// If for some reason was dead, we will recreate it.
+	if controller == nil {
+		// Global informer is based on the CachedResource type and we construct index based on the schema labels.
+		controllerCtx, cancel := context.WithCancel(ctx)
+
+		global := r.cacheKcpInformers.Cache().V1alpha1().PublishedObjects()
+
+		// Local informer is based on the specific types we want to replicate.
+		// TODO: use ClusterWithContext
+		local, err := r.discoveringDynamicKcpInformers.Cluster(cluster).ForResource(gvr)
+		if err != nil {
+			logger.Error(err, "Failed to get local informer for resource", "resource", gvr)
+			cancel()
+			return reconcileStatusStopAndRequeue, err
+		}
+		// TODO(mjudeikis): dynamic RestMapper!
+		replicated := &replicationcontroller.ReplicatedGVR{
+			Kind:   strings.TrimSuffix(strings.Title(gvr.Resource), "s"), //nolint:staticcheck
+			Local:  local.Informer(),
+			Global: global.Informer(),
+		}
+		replicationcontroller.InstallIndexers(replicated)
+		callback := func() {
+			r.callback(publishedResource)
+		}
+
+		c, err := replicationcontroller.NewController(
+			r.shardName,
+			r.dynamicCacheClient,
+			r.kcpCacheClient,
+			gvr,
+			replicated,
+			callback,
+			labels.SelectorFromSet(publishedResource.Spec.LabelSelector.MatchLabels),
+		)
+		if err != nil {
+			cancel()
+			return reconcileStatusContinue, err
+		}
+
+		go replicated.Local.Run(ctx.Done())
+		go replicated.Global.Run(ctx.Done())
+
+		if !cache.WaitForCacheSync(ctx.Done(), replicated.Local.HasSynced, replicated.Global.HasSynced) {
+			cancel()
+			return reconcileStatusContinue, fmt.Errorf("failed to wait for informers to sync")
+		}
+
+		go func() {
+			c.Start(controllerCtx, 1)
+		}()
+
+		r.controllerRegistry.register(controllerName, c, cancel)
+		if publishedResource.Status.Phase != cachev1alpha1.PublishedResourcePhaseDeleting {
+			conditions.MarkTrue(publishedResource, cachev1alpha1.ReplicationStarted)
+			publishedResource.Status.Phase = cachev1alpha1.PublishedResourcePhaseReady
+		}
+		return reconcileStatusStopAndRequeue, nil // Once controller is started, we requeue to check if we need to delete it.
+	} else {
+		controller.SetLabelSelector(labels.SelectorFromSet(publishedResource.Spec.LabelSelector.MatchLabels))
+	}
```

Comment:
```suggestion
	}
	
	controller.SetLabelSelector(labels.SelectorFromSet(publishedResource.Spec.LabelSelector.MatchLabels))
```

Also pedantic, but since the `if` block always returns the `else` block can be omitted.